V0.x

abhiksark opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduces ArrPy: a pure‑Python ndarray, dtype system, and array-creation helpers (zeros, ones, arange, linspace, eye, array). Adds basic 1D/2D indexing and public APIs for broadcasting, manipulation, ufuncs, linalg, and random (mostly placeholders).

* **Documentation**
  * Adds README and a comprehensive Development Guide.

* **Tests**
  * Adds pytest fixtures and skeleton test suites for arrays and dtypes.

* **Chores**
  * Adds packaging, pyproject, Makefile, benchmarks, dev requirements, and .gitignore updates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

abhiksark committed:
feat: Initial project structure for ArrPy v0.1 (Pure Python)

- Set up complete project structure for NumPy recreation
- Added core modules: arrpy, dtype, ufuncs, broadcasting, linalg, random
- Created comprehensive test framework with pytest fixtures
- Added benchmarking suite to compare with NumPy
- Configured project with setup.py, pyproject.toml, and makefile
- Added development dependencies and documentation
- All modules have skeleton implementations ready for development

This establishes the foundation for implementing a complete NumPy clone
in pure Python for educational purposes. The project will evolve through
three versions: v0.x (Pure Python), v1.x (Cython), v2.x (C++/PyBind11).

abhiksark committed:
feat: Implement Phase 1 - Core foundation of ArrPy

- Implemented basic arrpy class with flattened data storage
- Added shape, size, ndim, dtype, and strides properties
- Implemented array creation functions (zeros, ones, full, empty, arange, linspace, eye)
- Added basic indexing support for 1D and 2D arrays (integer and slice)
- Implemented dtype system with type inference
- Added __str__ and __repr__ methods for array display

All Phase 1 tests pass with NumPy-compatible behavior.

abhiksark committed:
Update arrpy/linalg.py

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

aniveshpandey commented on arrpy/indexing.py:
```diff
@@ -0,0 +1,83 @@
+    pass
+
+
+def compute_slice_indices(slice_obj, dim_size):
```
Since we only implemented slicing for 1D and 2D arrays, let's only implement this for 1D and 2D. And maybe later in further versions we could write if for higher dimensions.

aniveshpandey commented on benchmarks/bench_creation.py:
```diff
@@ -0,0 +1,141 @@
+"""
```
Let's also write benchmarks for comparing our own versions with each other, not just with numpy as benchmark!

aniveshpandey commented on tests/test_arrpy.py:
```diff
@@ -0,0 +1,176 @@
+"""
```
Could we also try writing tests for total memory consumption efficiency, concatenation, insertion, deletion here?

abhiksark committed:
feat: Complete ArrPy v1.0.0 Release

- Added comprehensive backend support with Python, Cython, and C implementations.
- Introduced a new Makefile for development tasks and streamlined setup processes.
- Implemented extensive array operations, including creation, manipulation, and advanced indexing.
- Enhanced performance with optimized C++ extensions and zero-copy operations.
- Established a robust testing framework and benchmarks to validate functionality and performance.
- Updated documentation and added contribution guidelines to support community involvement.

This release marks the culmination of the educational journey from pure Python to optimized C++, demonstrating the complete optimization path for array computing.

aniveshpandey reviewed (commented):
ISSUE_TEMPLATE is a real nice addition.
